### PR TITLE
Add Quantity formatting settings UI and UnitFormattingSettingsProvider interface to persist settings.

### PR DESCRIPTION
**Pull request by bsteinbk**
- Add Quantity formatting settings "page" that allow format overrides to be be set by QuantitiyType and Presentation Unit System. 
- Add UnitFormattingSettingsProvider interface that can be set in QuantityFormatter to persist settings.
- Add LocalUnitFormatProvider which is an implementation of UnitFormattingSettingsProvider which uses local storage to save and retrieve unit formatting overrides.
- Updated many alpha tags dealing with Quantity Formatting to beta so links to public documentation could be used.

### CODE REVIEW

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts [hunk 1]
```diff
@@ -237,21 +239,60 @@ export interface FormatterParserSpecsProvider {
...
 export interface QuantityFormatsChangedArgs {
   // string that represents the QuantityType that has been overriden or the overrides cleared.
   readonly quantityType: string;
 }
 
+/** Arguments sent to UnitFormattingSettingsProvider when overrides are changed.
+ * @beta
+ */
+export interface QuantityFormatOverridesChangedArgs {
+  // string that represents the QuantityType that has been overriden or the overrides cleared.
```
Comment (pmconne) [inline core/frontend/src/quantity-formatting/QuantityFormatter.ts:261]: Use `/** jsdoc comments */` or the documentation tools will ignore them. Same problem elsewhere, e.g. UnitFormattingSettingsProvider.

File: docs/changehistory/NextVersion.md [hunk 2]
```diff
@@ -5,7 +5,54 @@ publish: false
...
+    />
+  );
+}
+```
+
+These new classes are marked as beta in this release and are subject to minor modifications in future releases.
+
+### Add Settings Page to set Quantity Formatting Overrides
+
+The [QuantityFormatSettingsPanel]($ui-framework) component has been added to the @bentley/ui-framework package to provide the UI to set both the [PresentationUnitSystem] ($presentation-common) and formatting overrides in the [QuantityFormatter](imodeljs-frontend). This panel can be used in the new [SettingsContainer]($ui-core) UI component. The function `getQuantityFormatsSettingsManagerEntry` will return a [SettingsTabEntry]($ui-core) for use by the [SettingsManager]($ui-core). Below is an example of registering the `QuantityFormatSettingsPanel` with the `SettingsManager`.
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:31]: PresentationUnitSystem won't show up as a link with a space between ] and (.

**Review by pmconne — CHANGES_REQUESTED** [review]

**Review by pmconne — APPROVED** [review]
Gotta run extract-api again after fixing docs.

**Review by DanEastBentley — APPROVED** [review]

File: core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+import { FormatProps } from "@bentley/imodeljs-quantity";
+import { IModelApp } from "../IModelApp";
+import { IModelConnection } from "../IModelConnection";
+import { SelectedViewportChangedArgs } from "../ViewManager";
+import {
+  FormattingUnitSystemChangedArgs, OverrideFormatEntry, QuantityFormatOverridesChangedArgs, QuantityFormatter,
+  QuantityTypeKey, UnitFormattingSettingsProvider, UnitSystemKey,
+} from "./QuantityFormatter";
+
+/** This abstract class can monitor active imodel changes and update the QuantityFormatter overrides and the active unit system based on
```
Comment (calebmshafer) [inline core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts:18]: Found the wording of the docs a bit hard to follow. What do you think of something like this?

```md
This abstract class reacts to changes in the currently active iModel and updates the [[QuantityFormatter]] overrides and active unit system based on stored preferences.

In addition, monitors the [[QuantityFormatter]] for changes to format overrides and the active unit system and stores these changes.
```

Also, do we qualify what an "active iModel" is anywhere else? If not, maybe add a note about what it means in this context.

File: core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Utils
```
Comment (calebmshafer) [inline core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts:6]: Update to `Quantity Formatting` module.

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts [hunk 5]
```diff
@@ -22,7 +23,7 @@ import { BasicUnitsProvider } from "./BasicUnitsProvider";
  * "imperial" -> PresentationUnitSystem.BritishImperial ("british-imperial")
  * "usCustomary" -> PresentationUnitSystem.UsCustomary ("us-customary")
  * "usSurvey" -> PresentationUnitSystem.UsSurvey ("us-survey")
- * @alpha
+ * @beta
  */
 export type UnitSystemKey = "metric" | "imperial" | "usCustomary" | "usSurvey";
```
Comment (calebmshafer) [inline core/frontend/src/quantity-formatting/QuantityFormatter.ts:28]: Before we make this beta, do we want to align our naming with Presentation Unit Systems?

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts [hunk 6]
```diff
@@ -237,21 +239,60 @@ export interface FormatterParserSpecsProvider {
...
- * @alpha
+ * @beta
  */
 export interface QuantityFormatsChangedArgs {
-  // string that represents the QuantityType that has been overriden or the overrides cleared.
+  /** string that represents the QuantityType that has been overriden or the overrides cleared. */
   readonly quantityType: string;
 }
 
+/** Arguments sent to UnitFormattingSettingsProvider when overrides are changed.
```
Comment (calebmshafer) [inline core/frontend/src/quantity-formatting/QuantityFormatter.ts:257]: `UnitFormattingSettingsProvider` -> `[[UnitFormattingSettingsProvider]]`

File: core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ */
+
+import { OverrideFormatEntry, QuantityFormatter, QuantityTypeKey, UnitSystemKey } from "./QuantityFormatter";
+import { BaseUnitFormattingSettingsProvider } from "./BaseUnitFormattingSettingsProvider";
+
+/** Implementation of BaseUnitFormattingSettingsProvider that stores and retrieves data in local storage.
+ *  @beta
+ */
+export class LocalUnitFormatProvider extends BaseUnitFormattingSettingsProvider {
+  /** if `maintainOverridesPerIModel` is true the base class will set up listeners to monitor active iModel
```
Comment (calebmshafer) [inline core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts:16]: `if`/`If`

File: docs/changehistory/NextVersion.md [hunk 8]
```diff
@@ -5,7 +5,54 @@ publish: false
...
+    />
+  );
+}
+```
+
+These new classes are marked as beta in this release and are subject to minor modifications in future releases.
+
+### Add Settings Page to set Quantity Formatting Overrides
+
+The [QuantityFormatSettingsPanel]($ui-framework) component has been added to the @bentley/ui-framework package to provide the UI to set both the [PresentationUnitSystem]($presentation-common) and formatting overrides in the [QuantityFormatter](imodeljs-frontend). This panel can be used in the new [SettingsContainer]($ui-core) UI component. The function `getQuantityFormatsSettingsManagerEntry` will return a [SettingsTabEntry]($ui-core) for use by the [SettingsManager]($ui-core). Below is an example of registering the `QuantityFormatSettingsPanel` with the `SettingsManager`.
```
Comment (calebmshafer) [inline docs/changehistory/NextVersion.md:31]: `[QuantityFormatter](imodeljs-frontend)` is missing the `$`

**Review by calebmshafer — CHANGES_REQUESTED** [review]
@bsteinbk if we're going to update the Quantity Formatting code to be `beta` we should probably do the same for the quantity package itself.  It is intertwined with most of the public API for this.

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts (hunk 5 above)
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts:28]: So are you suggesting to use the name "PresentationUnitSystemKey", change the entries to exactly match the strings in PresentationUnitSystem enum or both?  I avoided adding a dependency on the Presentation package and using the enum directly to avoid a circular dependency but I could use the exact same strings, it would just be a breaking change... but I guess that is OK since it was alpha.

File: docs/changehistory/NextVersion.md (hunk 8 above)
Comment (bsteinbk) [reply docs/changehistory/NextVersion.md]: fixed

File: core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts (hunk 7 above)
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts]: fixed

File: core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts (hunk 4 above)
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts]: Changed to QuantityFormatting as no other module name contain a space.

File: core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts (hunk 3 above)
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts]: Updated.

File: core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Utils
```
[the remark this answers is not in the document]
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts]: Done

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts (hunk 5 above)
Comment (calebmshafer) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts:28]: Yes, I was thinking before it moves to beta this is probably the last chance to align the two better. Don't have a good suggestion at the moment, I'll review more closely tomorrow (have to review @ColinKerr's #902 as well).

It definitely doesn't make sense to add a presentation dependency. Maybe at the very least have the string values align even if it's not the same type.

Looping @grigasp in as well.

File: core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts (hunk 4 above)
Comment (calebmshafer) [reply core/frontend/src/quantity-formatting/LocalUnitFormatProvider.ts]: Sounds good forgot about that restriction.

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts (hunk 5 above)
Comment (grigasp) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts:28]: I think it's inevitable for presentation packages to depend on this package in the future, so definitely not worth adding a presentation-common dependency in here. Having the string values match would be helpful until presentation has that dependency, but afterwards I think we're going to deprecate `PresentationUnitSystem` and switch to `UnitSystemKey` or at least have some function that maps from one to the other.

Also I should mention that `PresentationUnitSystem` is not a really good name for what it stands for. IMO I should've named it `PresentationUnitSystemGroup`, because it stands for a group of unit systems.

Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts:28]: - My vote would be to not change the string values defined by UnitSystemKey, but possibly change the strings used in PresentationUnitSystem enum to match, especially since this would not cause any side effects for existing API consumers of either package. I remember looking at changing QuantityFormatter to use the strings from the PresentationUnitSystem enum once, and it caused other issues with the JSON that was being serialized because I had used the UnitSystemKey as the key in the persisted JSON object and JSON did not like a "-" in the name of a key.
- The name "Presentation Unit System" is really only exposed as a string in the UI because we needed a name the would not be confused with the "UnitSystem"  from the units schema, for the few people that actually look at the units schema.  I can change the default label to "Unit System Group" to be more accurate but exiting apps like it1 and it3 already use the label "Unit System" in their UI.  I thought adding "Presentation" to it would be less confusing for existing UI's to adopt.

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts (hunk 6 above)
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts]: fixed

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts (hunk 5 above)
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts:28]: @grigasp if you are not opposed I will make the following change.
```
/** @alpha */
export enum PresentationUnitSystem {
  Metric = "metric",
  BritishImperial = "imperial",
  UsCustomary = "usCustomary",
  UsSurvey = "usSurvey",
}
```

Comment (grigasp) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts:28]: No can do - these strings are sent to the backend and used by the addon. So the addon needs to be updated to handle both variants until we get a chance to bump RPC major version.

If you don't want to change `UnitSystemKey`, I'd say just leave them as is for now. I see `presentation-common` already has a dependency on `imodeljs-quantity` - for now I'll add mapping functions to map between these two types and will switch to using `UnitSystemKey` when the time is right.

**Review by NancyMcCallB — APPROVED** [review]

File: core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+      }
+    } else {
+      // setting a new override or set of overrides
+      const storedJson = await this.retrieve (typeKey);
+      const updatedFormat = {...storedJson, ...overrideEntry};
+      await this.store (typeKey, updatedFormat);
+    }
+  };
+
+  /** save UnitSystem for active Imodel */
```
Comment (calebmshafer) [inline core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts:69]: `Imodel` -> `iModel`

Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts]: fixed

File: core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+              }
+            }
+          });
+        }
+      }
+    }
+    return overrideFormatProps;
+  }
+
+  // serializes JSON object containing format overrides for a specific quantity type.
```
[the remark this answers is not in the document]
Comment (bsteinbk) [reply core/frontend/src/quantity-formatting/BaseUnitFormattingSettingsProvider.ts]: done

**Review by calebmshafer — APPROVED** [review]

File: core/frontend/src/quantity-formatting/QuantityFormatter.ts (hunk 5 above)
Comment (calebmshafer) [reply core/frontend/src/quantity-formatting/QuantityFormatter.ts:26]: Sounds good, I'll resolve this for now and we can circle back to it separately.
